### ISSUE TITLE
defining foreign keys from JSONB field

### ISSUE THREAD

**Issue by mitar** [hasura/graphql-engine#3462]
It seems I cannot define a foreign key in the console which would map from a field in a JSONB field to a primary key in another table. Only fields themselves are available, but no way to select a JSONB field.

**Comment by mitar** [comment]
GraphQL relationships also cannot be defined from a JSONB field. :-(

**Comment by tirumaraiselvan** [comment]
Hey @mitar 

You can accomplish this relationship via computed fields esp https://docs.hasura.io/1.0/graphql/manual/schema/computed-fields.html#table-computed-fields

I created a function like:

```
CREATE OR REPLACE FUNCTION public.fetch_user_jsons(user_row usertable)
 RETURNS SETOF jsontable
 LANGUAGE sql
 STABLE
AS $function$
  SELECT *
  FROM jsontable j
  WHERE
    j.myjson -> 'mykey' = to_jsonb(user_row.id);
$function$

```

And then add this as a computed field:

```
{
    "type":"add_computed_field",
    "args":{
        "table":{
            "name":"usertable",
            "schema":"public"
        },
        "name":"jsonRel",
        "definition":{
            "function":{
                "name":"fetch_user_jsons",
                "schema":"public"
            },
            "table_argument":"user_row"
        }
    }
```

And then query it just like a relationship:

```
query {
  user {
    id
    name
    jsonRel {
      myjson
      some_other_column_in_jsontable
    }    
  }
}
```

**Comment by mitar** [comment]
So you changed a title of the issue, but my issue was that this is not possible to define in the console/web interface. I cannot define a foreign key in the web interface (allowing to have constraints, like conflict propagation), nor GraphQL relation.

**Comment by mitar** [comment]
But nevertheless, thank you for this example how to do this manually.

**Comment by rikinsk** [comment]
@mitar it is not possible to have foreign keys using a field inside a JSONB field in Postgres

**Comment by mitar** [comment]
So not directly, but you can construct it manually: define an index on the field, and then define triggers which cascade/delete/prevent changes, like a foreign key would.

**Comment by dminkovsky** [comment]
Console support for computed fields is in 1.1 (currently beta)

Is the above-mentioned solution going to be the recommended approach for the near term?

**Comment by tirumaraiselvan** [comment]
@dminkovsky Yes. Since there is no simple way to define "foreign key"-like relationships with JSONB keys in Postgres, computed fields which return SETOF is the best pattern to use here.

**Comment by mitar** [comment]
There are functional indices you can define on PostgreSQL and you can define triggers to get behavior of foreign keys. It would be great if Hasura could do this out of the box.

**Comment by rikinsk** [comment]
Reopening this to allow discussion on @mitar 's suggestions

**Comment by nolandg** [comment]
Having quick, easy, and somewhat error and typo controlled management of JSONB pseudo-foreign key relationships in Hasura Console would be amazing. It opens the door to creating some very flexible CMS functionality using almost all Hasura-native functionality. I could create a CMS where my clients can create their own "file" fields that relate to a file table. Dynamic table fields and CMS-like functionality are possible now but very limited when it comes to joining tables--joins and relationships must be done manually. Having Hasura handle this would make it a customer CMS machine!

**Comment by farzd** [comment]
@tirumaraiselvan 
this isnt working for me

I have a table called stages, which has a match column [type jsonb] 
its an array of UUID's ```[ "e838be48-46a8-4992-abfa-c179e2114ef0", "32e998b1-8607-4ff1-8fea-33aff1fc8941" ]```

then i have a table called matches, which has these corresponding UUID's

```
CREATE OR REPLACE FUNCTION public.fetch_matches3(stages_row stages)
 RETURNS SETOF matches
 LANGUAGE sql
 STABLE
AS $function$
  SELECT *
  FROM matches j
  WHERE
    j.id IN (SELECT jsonb_array_elements(stages_row.matches))
$function$
```

error = postgres-error : operator does not exist: uuid = jsonb

**Comment by DoisKoh** [comment]
Try `json_array_elements_text` instead. Assuming j.id is of type UUID, cast it to UUID instead (::UUID).
